### PR TITLE
Add AES encrypted keys support for PKCS5 PBES2

### LINKED ISSUE #7038: Support AES in PKCS5
The PBES2 tests in `test_suite_pkcs5` all rely on [DES, which is deprecated](https://github.com/Mbed-TLS/mbedtls/issues/7024). The PKCS5 module does not support AES, but it's cryptographically agile, so it should be easy to add AES support.

The goal of this task is:

* Add the [AES-CBC OID from RFC 8018](https://www.rfc-editor.org/rfc/rfc8018#appendix-C) to the OID module.
* In `test_suite_pkcs5`, add test cases using AES, equivalent to what we have for DES.

### PR DESCRIPTION
**Pull request by zvolin**
## Description

This adds support for using pkcs5 pbes2 with the aes-cbc encryption schemes.
References #7024. Closes #7038.

The test data was created with openssl as follows:
```
openssl kdf -keylen $LEN -kdfopt digest:SHA256 -kdfopt iter:10000 -kdfopt pass:"PasswordPasswordPassword" -kdfopt hexsalt:0102030405060708 PBKDF2

openssl enc -$ENC -iv 2F904F75B47B48A618068D79BD9A826C -K $KEY -in $IN -e -out $OUT
```

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **backport** not needed because this is a new feature
- [x] **tests** provided for encryption and decryption

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

### CODE REVIEW

File: include/mbedtls/pkcs5.h [hunk 1]
```diff
@@ -59,7 +59,7 @@ extern "C" {
  * \param pwdlen     length of password
  * \param data       data to process
  * \param datalen    length of data
- * \param output     output buffer
+ * \param output     output buffer. must be of size \p datalen
```
Comment (zvolin) [inline include/mbedtls/pkcs5.h:62]: this is only true for decryption tho

Comment (gilles-peskine-arm) [reply include/mbedtls/pkcs5.h:62]: It's good to document this (it's really a bug that it wasn't documented), but we need the documentation to be correct! `datalen` is not correct.

Under the hood, `mbedtls_pkcs5_pbes2` calls `mbedtls_cipher_crypt` with the same buffer lengths, so the same constraint applies. From the documentation of `mbedtls_cipher_crypt` (I just changed the parameter names):

> This must be able to hold at least `datalen + block_size`. This must not be the same buffer as `input`.

This documentation is in fact overly pessimistic and I suspect that many users don't actually respect this constraint. In practice, the output buffer only needs to be large enough for the actual output. That's still larger than the input for encryption.

File: tests/suites/test_suite_pkcs5.data [hunk 2]
```diff
@@ -110,6 +110,22 @@ PBES2 Decrypt (OK)
 depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_DES_C:MBEDTLS_CIPHER_MODE_CBC
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pkcs5.data:110]: Please add the algorithm to this test case's description.

File: tests/suites/test_suite_pkcs5.data [hunk 3]
```diff
@@ -110,6 +110,22 @@ PBES2 Decrypt (OK)
...
...0A388D1AB7793004B83B97915D747F8CB09390D2C2A0DE05FCAFAA36911DA2F5D30E0E0178FA7B7B0E57006A44056169078690F9B48225340812A2ADA48CC56B48A0FC75C842DE94FA702818100D19D86F612C92312B3C7F673CA9F4C4808ABEBAD925526A4509E3179C025DFE351720B40562355955B5856A0C9B306F0AAA3C132CEF8E04D94651FCE6A6B89FAD6A683E40551743D962FB4AE5895F7186D9EEC1F5E8A11CCC8BE1783080C5C86E909CF366B26E79316AC687F42E6A6300BC50B944642CF297393FE9EF47F3CED0281806035C18ABC754C1610314177BA2947159A777E7D2F0903643289FD29C58694A3CB521AD7A97510989B0CDCA29B1A1E74632B84D16884DE13931356927391C7F9B9F1686D213A8D5D2620B1920FFEC1971399138D79013063DA8D3ABAC5B62CFAFD96B51DD0ACE4DD6F31440C47555AEBB849CC8B7A1A48526BD67B4547A98ABF0281800E8210AD476F9475DEEBA02B3383030C994D0CEADE8DF636B93DB9BEEB85B3915039CC7D8F78B29BBBCEC9702785B49D43B241D4FA6E4F580F45508640E7608D73FDE2730DCB2D7875C4A784C9ACD7856150F696D53C492F1B55CCDACCDBCE2BA8457D8A0EB08AA6AB9313487BA0EB2FB820D34457F897CE0EA9D79DB8DE75A102818100DB30BDD1E07CB8131091FB3DE7284553FCB0C5C9797C02A74D364D235A41DBCB3D210528722440AF91B073BD5D8EEE0123DC32FF67CE8410E9875B3061AA976E0470797A03C7A59B4A25986AF2C79D2237AABADEF8A9F7BD117CA7274FBFAB4DB7BB77583720419A14938FF1DC643D7AAF7119A3DC46F7F178778C51310579A1"
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pkcs5.data:115]: Where is this test data coming from? Please add the source in a comment. We used to be bad with this in the past and now we have some test data of unknown origin which we can't validate or recreate. But now we're careful, and we want to know the origin of all new test data.

Per my comment on `host_test.function`, if possible, we would prefer smaller data that fits in existing buffer sizes.

File: include/mbedtls/oid.h [hunk 4]
```diff
@@ -282,6 +282,9 @@
         MBEDTLS_OID_OIW_SECSIG_ALG "\x07"                                                                        /**< desCBC OBJECT IDENTIFIER ::= { iso(1) identified-organization(3) oiw(14) secsig(3) algorithms(2) 7 } */
```
Comment (gilles-peskine-arm) [inline include/mbedtls/oid.h:282]: Missing: since this is a new feature, it needs a [changelog entry](https://github.com/Mbed-TLS/mbedtls/blob/development/ChangeLog.d/00README.md).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
Thank you for your contribution! It needs a little more work.

File: include/mbedtls/pkcs5.h (hunk 1 above)
Comment (Ryan-Everett-arm) [reply include/mbedtls/pkcs5.h]: Merged work has now addressed this lack of documentation (https://github.com/Mbed-TLS/mbedtls/commit/708d78f80b8b3c3cda9f46977a67625149c30591)

**Comment by mpg** [comment]
This looks like it already has two reviewers, so I'm removing the label "needs-reviewer". Please correct me if that was wrong.

**Comment by mpg** [comment]
Ah no sorry just after writing see, I see @Ryan-Everett-arm pushed commits so he can no longer count as a reviewer.

File: include/mbedtls/oid.h [hunk 5]
```diff
@@ -306,6 +306,9 @@
...
 #define MBEDTLS_OID_DES_EDE3_CBC                MBEDTLS_OID_RSA_COMPANY "\x03\x07" /**< des-ede3-cbc OBJECT IDENTIFIER ::= { iso(1) member-body(2) -- us(840) rsadsi(113549) encryptionAlgorithm(3) 7 } */
 #define MBEDTLS_OID_AES                         MBEDTLS_OID_NIST_ALG "\x01" /** aes OBJECT IDENTIFIER ::= { joint-iso-itu-t(2) country(16) us(840) organization(1) gov(101) csor(3) nistAlgorithm(4) 1 } */
+#define MBEDTLS_OID_AES_128_CBC                 MBEDTLS_OID_AES "\x02" /** aes128-cbc-pad OBJECT IDENTIFIER ::= { joint-iso-itu-t(2) country(16) us(840) organization(1) gov(101) csor(3) nistAlgorithms(4) aes(1) aes128-CBC-PAD(2) } */
```
Comment (mpg) [inline include/mbedtls/oid.h:309]: Pre-existing: in this paragraph's opening comment ("Encryption algorithms") it would be good to add a reference to https://datatracker.ietf.org/doc/html/rfc8018#appendix-C

File: tests/suites/test_suite_pkcs5.data [hunk 6]
```diff
@@ -142,6 +142,30 @@ PBES2 Decrypt (Invalid padding & PKCS7 padding enabled)
 depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_DES_C:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_CIPHER_PADDING_PKCS7
 pbes2_decrypt:MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE:"301B06092A864886F70D01050C300E04082ED7F24A1D516DD702020800301406082A864886F70D030704088A4FCC9DCC394910":"70617373776f7264":"1B60098D4834CA752D37B430E70B7A085CFF86E21F4849F969DD1DF623342662443F8BD1252BF83CEF6917551B08EF55A69C8F2BFFC93BCB2DFE2E354DA28F896D1BD1BFB972A1251219A6EC7183B0A4CF2C4998449ED786CAE2138437289EB2203974000C38619DA57A4E685D29649284602BD1806131772DA11A682674DC22B2CF109128DDB7FDA3488A7144097565":144:MBEDTLS_ERR_PKCS5_PASSWORD_MISMATCH:"308187020100301306072A8648CE3D020106082A8648CE3D030107046D306B0201010420F12A1320760270A83CBFFD53F6031EF76A5D86C8A204F2C30CA9EBF51F0F0EA7A1440342000437CC56D976091E5A723EC7592DFF206EEE7CF9069174D0AD14B5F768225962924EE500D82311FFEA2FD2345D5D16BD8A88C26B770D55CD8A2A0EFA01C8B4EDFF060606060607"
 
+PBES2 Encrypt AES-128-CBC (OK)
```
Comment (mpg) [inline tests/suites/test_suite_pkcs5.data:145]: Minor: the provenance of the test data is documented in the commit message, but it would probably be good to have at least a hint directly in the file too. For example add "OpenSSL" in each test's description, and/or add a comment to the first test of the series explaining how the data was generated.

(Note: in the past we had lots of test data of undocumented provenance. Documenting it in the commit message is already good, but I'm trying to raise the bar even further.)

**Review by mpg — APPROVED** [review]
Looks good to me, just two minor nits.

One thing that may or may not be out of scope: I _think_ with the changes make here we're also getting support for PBES2 AES-encrypted key in PK parse. If that's the case, it would be nice to add a handlful of test cases demonstrating that in `test_suite_pkparse.data` (look for existing cases that depend on `PKCS5_C` for templates).

This could also be done in a follow-up, but since there's a chance it will Just Work™ thought I might suggest it here as a possible quick win. (However if generating test data turns out to be a lot of work, or the test don't pass without further work, then a follow-up is probably better.)

**Review by mpg — CHANGES_REQUESTED** [review]
There are CI failures. If I read it correctly, the line
```
depends.py: !MBEDTLS_SHA1_C MBEDTLS_SHA256_C FAILED
```
means that tests are failing in the two following configs:
- all hashes enabled except for SHA-1;
- only SHA-256 enabled (all other hashes disabled).

The declared dependencies for some of the new tests probably need updating.

**Comment by gilles-peskine-arm** [comment]
> with the changes make here we're also getting support for PBES2 AES-encrypted key in PK parse. If that's the case, it would be nice to add a handlful of test cases demonstrating that

Support in pk parse is the whole point of doing this. AFAIK that's the only reason we care about PBES2. So please do add at least a couple of integration test cases.

In the medium term (but not necessarily in this pull request), we want everything that's tested with DES to be also tested with AES, so that [if we remove DES, we won't lose test coverage](https://github.com/Mbed-TLS/mbedtls/issues/7024).

File: tests/suites/test_suite_pkcs5.data (hunk 6 above)
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pkcs5.data:145]: Counterpoint: a comment could get out of date if the test data changes, a commit message doesn't. So I prefer the commit message for documenting how the data was generated. A comment is useful for a slightly different reason: explaining how to generate similar data (for example, to generate a non-regression test case for a bug fix in a corner case).

Comment (mpg) [reply tests/suites/test_suite_pkcs5.data:145]: > Counterpoint: a comment could get out of date if the test data changes, a commit message doesn't.

Good point, agreed. I'd still like something like "(OpenSSL)" in the test case description as a hint that it wasn't taken from an RFC or other standard source of known-answers tests, but I won't insist on it.

File: tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_aes128_sha256.pem [hunk 7]
```diff
@@ -0,0 +1,30 @@
+-----BEGIN ENCRYPTED PRIVATE KEY-----
```
Comment (mpg) [inline tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_aes128_sha256.pem:1]: Thanks for mentioning how these were generated in the commit message. However for files in `tests/data_files` we have executable documentation in the form of a `Makefile`. The files are still committed normally, and this Makefile is not part of the build system, but it's the dedicated place to document how the files were generated.

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for addressing my feedback! Just one documentation point left.

Also, since this PR was created by a contributor without write permission, the CI's not starting automatically. After you push, can you start the CI manually by navigating to <https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/view/change-requests/>, logging in, looking for your PR and pressing the green arrow button on its line? (Same with pr-merge ideally.) (I've just done it for the current iteration.)

File: tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_aes128_sha256.pem (hunk 7 above)
Comment (Ryan-Everett-arm) [reply tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_aes128_sha256.pem]: Thanks! I didn't know about this Makefile. I have changed the test data to be more consistent with previous data generated, and added some DER tests as well.

Also, I have logged in to this page, but do not have any green arrows or other indication to start the CI.

File: ChangeLog.d/add-aes-cbc-to-pkcs5-pbes2.txt [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Features
+   * Add support for using AES-CBC 128, 192, and 256 bit schemes
+     with PKCS#5 PBES2.
```
Comment (mpg) [inline ChangeLog.d/add-aes-cbc-to-pkcs5-pbes2.txt:3]: Perhaps mention that this means keys encrypted this way can now be parsed by PK parse? That might not be obvious to all users.

File: tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_3des_sha384.pem [hunk 9]
```diff
@@ -1,30 +1,30 @@
...
-G9RvcdpVV8C2Y+M9qI2+x93WAtQ+NRJo4/+gJ0O9bVUhjjAmIHu2bMtbvr9aPJhd
-OpB9VQxB3c5mEXkNOV52oOGnIGVjbJMb4e3/MRpWtTFVcX6r200Gn6Hn3MnWZXdd
-H7pOpAowTcTlFcbJ0WWjfZygj5HKKUOFzPYNnXKizjzQhF6yK0mphKFY+8tpFQqB
-mV/1HlWJTSsAmh/FN21B2qq+KRiwMdpzKIEKC47mK+dzzo1mrTqmExvbiaLG8upr
-KMb/lEnSCasiZKTh71J3+5vUE+Nw73rYNZcdh7fj+GBK9KJ3hdKwYc/9yyQx1Lua
-4aXnUM6vQAsV+OLYNQE8vXMRtuftbPbV9sqiBLPIc/0P2EJ9mbEye8FM+koHUCKo
-xtJe5SK36DMwAas6tjimouVgWTcAdbq9r8jQlCJ1WxXPUcCJdv6pFQUGKQ+34TMK
-uWOhErUNRdqel9DthU5ig5dZs2DqlzbRzWYosZc1B6Q4/nua2JiBi8IeqtPILr2a
-JYJ9DNzxn07lcFHiVgrJuA==
+MIIFHDBOBgkqhkiG9w0BBQ0wQTApBgkqhkiG9w0BBQwwHAQI11AY9mDsg8QCAggA
```
Comment (mpg) [inline tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_3des_sha384.pem:2]: I don't think modifying this file was necessary, but OTOH I don't think it hurts either.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks almost entirely good to me, just requesting a clarification in the ChangeLog entry.

File: tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_3des_sha384.pem (hunk 9 above)
Comment (Ryan-Everett-arm) [reply tests/data_files/rsa_pkcs8_pbes2_pbkdf2_2048_3des_sha384.pem]: Apologies, I regenerated this file while investigating the data_files test generation and accidentally committed it. I have now reverted the change.

**Review by mpg — APPROVED** [review]
Looks all good to me now! Thank you for the quick rework every time!

File: tests/suites/test_suite_pkcs5.data [hunk 10]
```diff
@@ -142,6 +142,30 @@ PBES2 Decrypt (Invalid padding & PKCS7 padding enabled)
...
+
+PBES2 Encrypt AES-256-CBC (OK, generated with OpenSSL)
+depends_on:MBEDTLS_MD_CAN_SHA256:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC:!MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH
+pbes2_encrypt:MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE:"302a06092A864886F70D01050C301d0408010203040506070802022710020118300a06082A864886F70D0209301D060960864801650304012A04102F904F75B47B48A618068D79BD9A826C":"50617373776F726450617373776F726450617373776F7264":"5468697320697320612066696c6520746f20626520656e6372797074656420776974682050424553322c20776974682061206b65792067656e657261746564207573696e67206120707266202853484132353629":96:0:"7077e6e8e679962c6feda69c704d58507d143cf77be284f773dfcaa4a5f5e7c1cd5527204916302d435071c01126f4aa76205ce61998d7b8fdf065a77d4a77634376b9968e4722c27f2ac85d79565ff4fca9204d3e4bc8c5bd53c1785bb6e566"
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pkcs5.data:155]: Info: instantiating the commands in 223e716693f969561db05d887e66b6aad576e814 for the data in the test cases:
```
<<<5468697320697320612066696c6520746f20626520656e6372797074656420776974682050424553322c20776974682061206b65792067656e657261746564207573696e67206120707266202853484132353629 xxd -r -p | openssl enc -aes-128-cbc -iv 2f904f75b47b48a618068d79bd9a826c -K $(openssl kdf -keylen 16 -kdfopt digest:SHA1 -kdfopt iter:10000 -kdfopt pass:"PasswordPasswordPassword" -kdfopt hexsalt:0102030405060708 PBKDF2 | tr -d :) -e | xxd -p -c256
<<<5468697320697320612066696c6520746f20626520656e6372797074656420776974682050424553322c20776974682061206b65792067656e657261746564207573696e67206120707266202853484132353629 xxd -r -p | openssl enc -aes-192-cbc -iv 2f904f75b47b48a618068d79bd9a826c -K $(openssl kdf -keylen 24 -kdfopt digest:SHA256 -kdfopt iter:10000 -kdfopt pass:"PasswordPasswordPassword" -kdfopt hexsalt:0102030405060708 PBKDF2 | tr -d :) -e | xxd -p -c256
7c648a9df9759ba49283c261269a71bc06d45f6c24ab6431d77b2ecec1fd6d1aa751bd05b1c026fc8ff91baeb1b1838aa0f825b23be79bc09331d0607181e234dfea4ab3cbf7997747516486b6865e85de95dc9b64d45462197c891b31af6c94
<<<5468697320697320612066696c6520746f20626520656e6372797074656420776974682050424553322c20776974682061206b65792067656e657261746564207573696e67206120707266202853484132353629 xxd -r -p | openssl enc -aes-256-cbc -iv 2f904f75b47b48a618068d79bd9a826c -K $(openssl kdf -keylen 32 -kdfopt digest:SHA256 -kdfopt iter:10000 -kdfopt pass:"PasswordPasswordPassword" -kdfopt hexsalt:0102030405060708 PBKDF2 | tr -d :) -e | xxd -p -c256
7077e6e8e679962c6feda69c704d58507d143cf77be284f773dfcaa4a5f5e7c1cd5527204916302d435071c01126f4aa76205ce61998d7b8fdf065a77d4a77634376b9968e4722c27f2ac85d79565ff4fca9204d3e4bc8c5bd53c1785bb6e566
```

File: tests/suites/test_suite_pkcs5.data [hunk 11]
```diff
@@ -142,6 +142,30 @@ PBES2 Decrypt (Invalid padding & PKCS7 padding enabled)
 depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_DES_C:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_CIPHER_PADDING_PKCS7
 pbes2_decrypt:MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE:"301B06092A864886F70D01050C300E04082ED7F24A1D516DD702020800301406082A864886F70D030704088A4FCC9DCC394910":"70617373776f7264":"1B60098D4834CA752D37B430E70B7A085CFF86E21F4849F969DD1DF623342662443F8BD1252BF83CEF6917551B08EF55A69C8F2BFFC93BCB2DFE2E354DA28F896D1BD1BFB972A1251219A6EC7183B0A4CF2C4998449ED786CAE2138437289EB2203974000C38619DA57A4E685D29649284602BD1806131772DA11A682674DC22B2CF109128DDB7FDA3488A7144097565":144:MBEDTLS_ERR_PKCS5_PASSWORD_MISMATCH:"308187020100301306072A8648CE3D020106082A8648CE3D030107046D306B0201010420F12A1320760270A83CBFFD53F6031EF76A5D86C8A204F2C30CA9EBF51F0F0EA7A1440342000437CC56D976091E5A723EC7592DFF206EEE7CF9069174D0AD14B5F768225962924EE500D82311FFEA2FD2345D5D16BD8A88C26B770D55CD8A2A0EFA01C8B4EDFF060606060607"
 
+PBES2 Encrypt AES-128-CBC (OK, generated with OpenSSL)
+depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pkcs5.data:146]: Missing dependency: when encrypting (not when decrypting), `MBEDTLS_CIPHER_PADDING_PKCS7` is also needed.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from some missing depends and a copypasta.

File: tests/suites/test_suite_pkcs5.data [hunk 12]
```diff
@@ -143,16 +143,16 @@ depends_on:MBEDTLS_MD_CAN_SHA1:MBEDTLS_DES_C:MBEDTLS_CIPHER_MODE_CBC:MBEDTLS_CIP
...
+depends_on:MBEDTLS_MD_CAN_SHA256:MBEDTLS_AES_C:MBEDTLS_CIPHER_MODE_CBC:!MBEDTLS_AES_ONLY_128_BIT_KEY_LENGTH:MBEDTLS_CIPHER_PADDING_PKCS7
+pbes2_encrypt:MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE:"302706092A864886F70D01050C301a0408010203040506070802022710300a06082A864886F70D0209301D060960864801650304012A04102F904F75B47B48A618068D79BD9A826C":"50617373776F726450617373776F726450617373776F7264":"5468697320697320612066696c6520746f20626520656e6372797074656420776974682050424553322c20776974682061206b65792067656e657261746564207573696e67206120707266202853484132353629":96:0:"7077e6e8e679962c6feda69c704d58507d143cf77be284f773dfcaa4a5f5e7c1cd5527204916302d435071c01126f4aa76205ce61998d7b8fdf065a77d4a77634376b9968e4722c27f2ac85d79565ff4fca9204d3e4bc8c5bd53c1785bb6e566"
```
Comment (mpg) [inline tests/suites/test_suite_pkcs5.data:155]: Note: ideally this would have been two commits: one for the dependencies, one for the test data. Absolutely no need to rewrite history for this, just pointing it out for the future.

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Comment by gilles-peskine-arm** [comment]
The CI on the merge queue failed. [`test_suite_pkparse` is failing in some configurations](https://mbedtls.trustedfirmware.org/blue/organizations/jenkins/mbed-tls-pr-head/detail/gh-readonly-queue%2Fdevelopment%2Fpr-7604-8bdd8cdc4fc99588f6e3ed47d3dac7b2ddbdb6af/1/pipeline/485). Maybe a dependency needs to change?

**Comment by mpg** [comment]
I think the new tests need to depend on `CIPHER_C` as well. Actually, all test cases in `test_suite_pkparse.data` that already depend on `PKCS5_C` or `PKCS12_C` should now also depend on `CIPHER_C`. That's because we recently moved from:
- `PKCS5_C` and `PKCS12_C` each depend on `CIPHER_C`, to
- `PKCS5_C` and `PKCS12_C` no longer hard-depend on `CIPHER_C`, but only the en/decryption functions do.

So, "I want PKCS#5 (resp. 12) encryption functions" is now spelled `PKCS5_C && CIPHER_C` (resp. `PKCS12_C && CIPHER_C`).

**Review by gilles-peskine-arm — DISMISSED** [review]
LGTM

**Review by mpg — APPROVED** [review]
LGTM

**Comment by gilles-peskine-arm** [comment]
The merge queue failed again, this time in [outcome analysis](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/gh-readonly-queue%2Fdevelopment%2Fpr-7604-c8de36220298f67254c253d274826d77f1022c34/1/execution/node/8376/log/). There's already an exception for some PKCS5 and PKCS12 tests in `analyze_driver_vs_reference_cipher_aead_cmac`, and with the latest changes in `development` I think we need to add the same exception to `analyze_block_cipher_dispatch`.

**Comment by gilles-peskine-arm** [comment]
@Ryan-Everett-arm At this stage, a merge is probably easier, especially since there are no conflicts.

**Comment by Ryan-Everett-arm** [comment]
> The merge queue failed again, this time in [outcome analysis](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/gh-readonly-queue%2Fdevelopment%2Fpr-7604-c8de36220298f67254c253d274826d77f1022c34/1/execution/node/8376/log/). There's already an exception for some PKCS5 and PKCS12 tests in `analyze_driver_vs_reference_cipher_aead_cmac`, and with the latest changes in `development` I think we need to add the same exception to `analyze_block_cipher_dispatch`.

The exceptions for pkcs5 are `PBES2 Encrypt, .*` and `PBES2 Decrypt .*`.  The discrepancy with the comma means my new encryption tests won't be matched here. Should I add a comma to my encryption tests to match the regex, or change the regex to `PBES2 Encrypt.*`?

**Comment by gilles-peskine-arm** [comment]
> Should I add a comma to my encryption tests to match the regex, or change the regex to PBES2 Encrypt.*?

Either way, I don't think it matters.

**Review by gilles-peskine-arm — DISMISSED** [review]
LGTM

**Review by mpg — APPROVED** [review]

**Comment by gilles-peskine-arm** [comment]
[Outcome analysis is now complaining that the ignore pattern is too broad](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-7604-head/20/execution/node/8374/log/).

Tip: since the problem is purely in outcome analysis, you can download the `outcomes.csv.xz` artifact and run `analyze_outcomes.py` locally.

**Comment by mpg** [comment]
Ah, we got this wrong: this is not that much like `analyze_driver_vs_reference_cipher_aead_cmac`, where `CIPHER_C` is disabled when drivers are in use, hence no PKCS#5/12 en/decryption works at all in the driver component.

Here `CIPHER_C` is enabled in both components (reference and driver). The difference is that `AES_C` is not enabled in the driver component, and PKCS#5 is not expected to be able to use a driver for AES because [we explicitly decided it was out of scope for now](https://github.com/Mbed-TLS/mbedtls/issues/8337). So, the correct pattern should be `.*AES.*` - and for `test_suite_pkcs5` only. (The old tests will pass in the driver component because DES is not accelerated there - only AES, Aria and Camellia.)

In restrospect the answer is in the PR's title: it's adding support for AES-based PKCS5 encryption. That's not supported when AES is driver-only, so we need to exclude just that from the comparison: `.*AES.*` in `test_suite_pkcs5`. We should have noticed that `test_suite_pkcs12` was off-topic. (Or if we want to be specific, `PBES2 .* AES-.*` or something like that.)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by mpg — APPROVED** [review]
